File: code/FSR/bridge.R
# Layer-2 R surface for spec 009. The actual MATLAB/FSDA call stays in the
# Python bridge; this file owns reticulate setup and R-side shape checks. It is
# the `FSR` sibling of the spec-003/006 R surfaces.

.bridge_dir = local({
  # Prefer source(".../bridge.R") metadata, then fall back to common working
  # directories used by Rscript and interactive sessions.
  frames = sys.frames()
  for (i in rev(seq_along(frames))) {
    ofile = frames[[i]]$ofile
    if (!is.null(ofile) && nzchar(ofile) && basename(ofile) == "bridge.R") {
      return(normalizePath(dirname(ofile), winslash = "/", mustWork = TRUE))
    }
  }

  here = normalizePath(getwd(), winslash = "/", mustWork = TRUE)
  candidates = c(here, file.path(here, "code", "FSR"))
  for (candidate in candidates) {
    if (file.exists(file.path(candidate, "bridge.py"))) {
      return(normalizePath(candidate, winslash = "/", mustWork = TRUE))
    }
  }

  stop("Cannot locate code/FSR/bridge.py; source bridge.R from the repo or target directory.")
})

.require_reticulate = function() {
  if (!requireNamespace("reticulate", quietly = TRUE)) {
    stop("R package 'reticulate' is required for spec 009.")
  }
  asNamespace("reticulate")
}

.resolve_python = function(python) {
  # No machine-specific default in the repo: prefer FSDA_DEV_VENV (passed in as
  # `python`), then the active interpreter on PATH (python or python3, so macOS
  # works too), else stop with guidance.
  if (is.null(python) || !nzchar(python)) {
    found = Sys.which(c("python", "python3"))
    found = found[nzchar(found)]
    python = if (length(found) > 0) unname(found[[1]]) else ""
  }
  if (!nzchar(python)) {
    stop(
      "No Python interpreter found. Set FSDA_DEV_VENV to your venv's python ",
      "executable (Scripts\\python.exe on Windows, bin/python on macOS) that ",
      "has matlabengine installed."
    )
  }
  normalizePath(python, winslash = "/", mustWork = FALSE)
}

.configure_python = function(reticulate, python) {
  # Accept a direct executable, a virtualenv root, or a conda-style root. This
  # must run before any Python import because reticulate locks the interpreter.
  root_python = file.path(python, "python.exe")
  scripts_python = file.path(python, "Scripts", "python.exe")
  bin_python = file.path(python, "bin", "python")
  pyvenv_cfg = file.path(python, "pyvenv.cfg")

  if (file.exists(python)) {
    reticulate$use_python(python, required = TRUE)
  } else if (dir.exists(python) && file.exists(pyvenv_cfg)) {
    reticulate$use_virtualenv(python, required = TRUE)
  } else if (file.exists(root_python)) {
    reticulate$use_python(root_python, required = TRUE)
  } else if (file.exists(scripts_python)) {
    reticulate$use_python(scripts_python, required = TRUE)
  } else if (file.exists(bin_python)) {
    reticulate$use_python(bin_python, required = TRUE)
  } else {
    stop(
      paste0(
        "Python venv or executable not found: ", python,
        ". Set FSDA_DEV_VENV to the project venv or a Python executable with matlabengine."
      )
    )
  }
}

.as_double_matrix = function(x, name) {
  # Strip dimnames and coerce numeric storage while preserving the explicit
  # matrix shape reticulate will pass to Python.
  if (!is.matrix(x) || length(dim(x)) != 2 || !is.numeric(x)) {
    stop(name, " must be a 2D numeric matrix.")
  }
  out = matrix(as.numeric(x), nrow = nrow(x), ncol = ncol(x))
  dimnames(out) = NULL
  out
}

.validate_inputs = function(y, X) {
  # Fail before crossing the R/Python boundary; bridge.py performs the matching
  # authoritative checks before calling MATLAB.
  if (is.matrix(y)) {
    if (ncol(y) != 1) {
      stop("y must be a numeric vector or n x 1 matrix.")
    }
    y = as.numeric(y)
  } else if (is.numeric(y) && is.null(dim(y))) {
    y = as.numeric(y)
  } else {
    stop("y must be a numeric vector or n x 1 matrix.")
  }
  n = length(y)
  if (n < 1) {
    stop("y must have at least one element.")
  }

  X = .as_double_matrix(X, "X")
  if (nrow(X) != n) {
    stop("X must have ", n, " rows to match y, got ", nrow(X), ".")
  }

  list(y = y, X = X)
}

.validate_bridge = function(bridge) {
  if (!inherits(bridge, "fsda_fsr_bridge") ||
      is.null(bridge$module) ||
      is.null(bridge$engine)) {
    stop("bridge must be a handle returned by start_bridge().")
  }
}

.scalar_string = function(x) {
  paste(as.character(unlist(x, recursive = TRUE, use.names = FALSE)), collapse = " ")
}

start_bridge = function(python = Sys.getenv("FSDA_DEV_VENV"), fsda_root = NULL) {
  # Import the local Python bridge and start a reusable MATLAB engine session.
  reticulate = .require_reticulate()
  python = .resolve_python(python)
  .configure_python(reticulate, python)

  # Python caches imported modules by their bare name, and every FSDA target's
  # Layer-1 file is named bridge.py. Evict any 'bridge' cached by a different
  # target (e.g. mahalFS / Score) so THIS target's code/FSR/bridge.py is loaded
  # fresh. Without this, using two targets in one R session returns the first-
  # imported module (AttributeError on the other target's helpers) until R restart.
  reticulate$py_run_string("import sys; sys.modules.pop('bridge', None)")
  module = reticulate$import_from_path("bridge", path = .bridge_dir, convert = TRUE)
  if (is.null(fsda_root) || !nzchar(fsda_root)) {
    engine = module$start_engine()
  } else {
    engine = module$start_engine(fsda_root = fsda_root)
  }

  handle = list(
    module = module,
    engine = engine,
    python = python,
    bridge_dir = .bridge_dir
  )
  class(handle) = "fsda_fsr_bridge"
  handle
}

fsr = function(bridge, y, X, nsamp = 0, intercept = TRUE, plots = 0, msg = 0,
               h = NULL, init = NULL, bonflev = NULL) {
  # Run FSDA FSR through the Python bridge; return the key fields as plain R.
  .validate_bridge(bridge)
  inputs = .validate_inputs(y, X)
  # reticulate maps R named args to Python keyword args (NULL -> None); bridge.py
  # reads the struct fields, normalizes the 1-based outlier indices
  # (scalar/array/empty) to a clean vector, and handles plots/msg (default off;
  # msg routes MATLAB's messages to this console, plots opens live figure windows).
  res = bridge$module$fsr(
    bridge$engine, inputs$y, inputs$X,
    nsamp = as.integer(nsamp), intercept = isTRUE(intercept),
    plots = as.integer(plots), msg = as.integer(msg),
    h = h, init = init, bonflev = bonflev
  )

  mdr = as.matrix(res$mdr)
  if (ncol(mdr) != 2) {
    stop("expected FSR mdr to have 2 columns, got ", ncol(mdr))
  }
  dimnames(mdr) = NULL
  list(
    mdr = mdr,
    outliers = as.integer(res$outliers),   # 1-based unit indices (empty if none)
    beta = as.numeric(res$beta),
    scale = as.numeric(res$scale)
  )
}

render_figures = function(bridge) {
  # Force any open MATLAB figures (from fsr(plots=...)) to paint.
  .validate_bridge(bridge)
  bridge$module$render_figures(bridge$engine)
  invisible(NULL)
}

wait_for_figures = function(bridge) {
  # Block until the user closes all open MATLAB figures. Driven MATLAB-side
  # (uiwait), so it is immune to the terminal-stdin interference that makes
  # reading a key in R fail once the engine is embedded via reticulate. Returns
  # at once if no figures are open.
  .validate_bridge(bridge)
  bridge$module$wait_for_figures(bridge$engine)
  invisible(NULL)
}

stop_bridge = function(bridge) {
  # MATLAB engine startup is expensive, so callers control shutdown explicitly.
  .validate_bridge(bridge)
  bridge$module$stop_engine(bridge$engine)
  invisible(NULL)
}

bridge_diagnostics = function(bridge) {
  # Keep MATLAB-specific diagnostics behind Python helper functions to avoid
  # reticulate probing MATLAB engine method signatures directly.
  .validate_bridge(bridge)
  reticulate = .require_reticulate()
  py_config = reticulate$py_config()
  metadata = reticulate$import("importlib.metadata", convert = TRUE)
  engine_pkg = tryCatch(
    metadata$version("matlabengine"),
    error = function(e) "n/a"
  )

  list(
    r = as.character(getRversion()),
    reticulate = as.character(utils::packageVersion("reticulate")),
    python = .scalar_string(py_config$python),
    python_version = .scalar_string(py_config$version_string),
    matlab = bridge$module$matlab_version(bridge$engine),
    matlabengine = engine_pkg,
    fsr_path = bridge$module$which_fsr(bridge$engine),
    bridge_dir = bridge$bridge_dir,
    requested_python = bridge$python
  )
}

File: code/FSR/bridge.jl
# Layer-2 Julia surface for spec 008. The actual MATLAB/FSDA call stays in the
# Python bridge (code/FSR/bridge.py); this file owns PythonCall setup and
# Julia-side shape checks. It mirrors bridge.R (spec 009) one-for-one:
# start_bridge / fsr / stop_bridge / bridge_diagnostics over an opaque handle
# that bundles the imported Python `bridge` module and the live engine. It is the
# `FSR` sibling of the spec-002/005 Julia surfaces.
#
# CRITICAL: PythonCall binds its interpreter at load time (`using PythonCall`),
# so the interpreter is resolved and exported via ENV *before* that line — there
# is no second chance. One-shot caveat: if PythonCall was already initialised in
# this Julia session against a different interpreter it cannot be rebound; start
# a fresh `julia` process to switch interpreters.

# --- Locate this file's directory (bridge.py must sit next to it) ------------
const _BRIDGE_DIR = let
    here = @__DIR__
    if isfile(joinpath(here, "bridge.py"))
        abspath(here)
    else
        # Fall back to common working dirs (repo root or the target folder).
        wd = pwd()
        candidates = [wd, joinpath(wd, "code", "FSR")]
        idx = findfirst(c -> isfile(joinpath(c, "bridge.py")), candidates)
        idx === nothing && error(
            "Cannot locate code/FSR/bridge.py next to bridge.jl or under the " *
            "working directory; run from the repo or the target directory.")
        abspath(candidates[idx])
    end
end

# --- Interpreter resolution (no machine-specific path committed) -------------
function _resolve_python(python::AbstractString)
    # Mirror bridge.R's precedence: an explicit value (FSDA_DEV_VENV) wins, else
    # the active python/python3 on PATH (so macOS works too), else a clear error.
    # An explicit value may be a python executable, a virtualenv root, or a
    # conda-style root; turn whatever we are given into a concrete executable.
    p = String(python)
    if isempty(p)
        found = Sys.which("python")
        found === nothing && (found = Sys.which("python3"))
        found === nothing && error(
            "No Python interpreter found. Set FSDA_DEV_VENV to your venv's python " *
            "executable (Scripts\\python.exe on Windows, bin/python on macOS) that " *
            "has matlabengine installed.")
        return abspath(found)
    end
    isfile(p) && return abspath(p)               # direct executable
    if isdir(p)                                  # venv / conda root
        nested = [
            joinpath(p, "python.exe"),            # conda root (Windows)
            joinpath(p, "Scripts", "python.exe"), # venv (Windows)
            joinpath(p, "bin", "python"),         # venv / conda (POSIX)
            joinpath(p, "bin", "python3"),
        ]
        idx = findfirst(isfile, nested)
        idx !== nothing && return abspath(nested[idx])
    end
    error("Python venv or executable not found: ", p,
          ". Set FSDA_DEV_VENV to the project venv or a Python executable with matlabengine.")
end

# Resolve the interpreter and pin PythonCall to it BEFORE `using PythonCall`.
const _PYTHON_EXE = _resolve_python(get(ENV, "FSDA_DEV_VENV", ""))
ENV["JULIA_CONDAPKG_BACKEND"] = "Null"   # do not let CondaPkg manage a Python
ENV["JULIA_PYTHONCALL_EXE"] = _PYTHON_EXE

using PythonCall

const _PYTHONCALL_VERSION = try
    string(pkgversion(PythonCall))
catch
    "n/a"
end

# --- Opaque handle (the bridge.R `fsda_fsr_bridge` list, as a struct) ---------
struct FSRBridge
    module_::Py     # imported Python `bridge` module
    engine::Py      # live matlab.engine session
    python::String  # interpreter PythonCall is bound to
    bridge_dir::String
end

# --- Julia-side input validation (fail before crossing into Python) ----------
function _validate_inputs(y, X)
    # bridge.py performs the matching authoritative checks before calling MATLAB;
    # these guards just give a Julia-side error before the boundary crossing.
    yv = if y isa AbstractVector
        Vector{Float64}(y)
    elseif y isa AbstractMatrix && size(y, 2) == 1
        Vector{Float64}(vec(y))
    else
        error("y must be a numeric vector or n x 1 matrix.")
    end
    n = length(yv)
    n >= 1 || error("y must have at least one element.")

    (X isa AbstractMatrix) || error("X must be a 2D numeric matrix.")
    Xm = Matrix{Float64}(X)
    size(Xm, 1) == n ||
        error("X must have ", n, " rows to match y, got ", size(Xm, 1), ".")

    return (y = yv, X = Xm)
end

function _validate_bridge(bridge)
    (bridge isa FSRBridge) ||
        error("bridge must be a handle returned by start_bridge().")
end

# --- Lifecycle ---------------------------------------------------------------
"""
    start_bridge(; python = get(ENV, "FSDA_DEV_VENV", ""), fsda_root = nothing)

Import the local Python `bridge` and start a reusable MATLAB engine session,
returning an opaque `FSRBridge` handle. PythonCall is already bound to the
interpreter resolved at load time; passing a different `python` here cannot
rebind it, so it only warns and records what was actually used.
"""
function start_bridge(; python::AbstractString = get(ENV, "FSDA_DEV_VENV", ""),
                       fsda_root = nothing)
    requested = isempty(String(python)) ? _PYTHON_EXE : _resolve_python(python)
    if requested != _PYTHON_EXE
        @warn("PythonCall is already bound to a different interpreter; the " *
              "`python` argument cannot rebind it in this session. Start a fresh " *
              "`julia` process to switch interpreters.",
              bound = _PYTHON_EXE, requested = requested)
    end

    # Python caches imported modules by their bare name, and every FSDA target's
    # Layer-1 file is named bridge.py. Evict any 'bridge' cached by a different
    # target (e.g. mahalFS / Score) and force THIS target's dir to the front of
    # sys.path, so code/FSR/bridge.py is (re)loaded fresh. Without this, using two
    # targets in one Julia session returns the first-imported module (AttributeError
    # on the other target's helpers) — the PythonCall analogue of the reticulate fix.
    sys = pyimport("sys")
    sys.modules.pop("bridge", nothing)
    while pyconvert(Bool, sys.path.__contains__(_BRIDGE_DIR))
        sys.path.remove(_BRIDGE_DIR)   # drop stale duplicates of this dir
    end
    sys.path.insert(0, _BRIDGE_DIR)    # this target's bridge.py wins
    module_ = pyimport("bridge")

    engine = if fsda_root === nothing || isempty(String(fsda_root))
        module_.start_engine()
    else
        module_.start_engine(fsda_root = String(fsda_root))
    end

    return FSRBridge(module_, engine, _PYTHON_EXE, _BRIDGE_DIR)
end

"""
    fsr(bridge, y, X; nsamp=0, intercept=true, plots=0, msg=0,
        h=nothing, init=nothing, bonflev=nothing) -> NamedTuple

Run FSDA `FSR` through the Python bridge and return the key fields converted to
Julia: `mdr` (m×2 `Matrix{Float64}` of [step, min deletion residual]), `outliers`
(1-based `Vector{Int}`, empty if none), `beta` (`Vector{Float64}`), `scale`
(`Float64`). `nsamp=0` enumerates all subsets (deterministic). `plots`/`msg`
default to off; set them to open live MATLAB figure windows (keep the bridge
alive — see `render_figures`) and route FSR's messages to this terminal.
"""
function fsr(bridge, y, X; nsamp::Integer = 0, intercept::Bool = true,
             plots::Integer = 0, msg::Integer = 0,
             h = nothing, init = nothing, bonflev = nothing)
    _validate_bridge(bridge)
    inp = _validate_inputs(y, X)
    # PythonCall passes the Julia vector/matrix to Python keeping their logical
    # shape; bridge.py's guards (shapes, struct-field reads, 1-based outlier
    # normalization) and its plots/msg handling are authoritative. nothing -> None.
    res = bridge.module_.fsr(bridge.engine, inp.y, inp.X;
                             nsamp = nsamp, intercept = intercept,
                             plots = plots, msg = msg,
                             h = h, init = init, bonflev = bonflev)
    # PythonCall does NOT auto-convert; convert each field explicitly. No reshape.
    mdr = pyconvert(Matrix{Float64}, res["mdr"])
    (ndims(mdr) == 2 && size(mdr, 2) == 2) ||
        error("expected FSR mdr to be (m, 2), got ", size(mdr))
    return (
        mdr = mdr,
        outliers = pyconvert(Vector{Int}, res["outliers"]),
        beta = pyconvert(Vector{Float64}, res["beta"]),
        scale = pyconvert(Float64, res["scale"]),
    )
end

"""
    render_figures(bridge)

Force any open MATLAB figures (from `fsr(...; plots=…)`) to paint. Figures close
when the engine quits, so keep the bridge alive (don't `stop_bridge`) until you
have viewed them.
"""
function render_figures(bridge)
    _validate_bridge(bridge)
    bridge.module_.render_figures(bridge.engine)
    return nothing
end

"""
    wait_for_figures(bridge)

Block until the user closes all open MATLAB figures. Driven MATLAB-side (`uiwait`),
so it is immune to the terminal-stdin interference under PythonCall (reading a key
from Julia does not work once the engine is embedded). Returns at once if no
figures are open.
"""
function wait_for_figures(bridge)
    _validate_bridge(bridge)
    bridge.module_.wait_for_figures(bridge.engine)
    return nothing
end

"""
    stop_bridge(bridge)

Quit the MATLAB engine (startup is expensive, so callers control shutdown).
"""
function stop_bridge(bridge)
    _validate_bridge(bridge)
    bridge.module_.stop_engine(bridge.engine)
    return nothing
end

"""
    bridge_diagnostics(bridge) -> NamedTuple

Julia / PythonCall / Python / MATLAB / matlabengine / FSR-path details. MATLAB
values go through the Python helpers `bridge.matlab_version(eng)` and
`bridge.which_fsr(eng)` so PythonCall never probes engine method signatures.
"""
function bridge_diagnostics(bridge)
    _validate_bridge(bridge)
    sys = pyimport("sys")
    metadata = pyimport("importlib.metadata")
    engine_pkg = try
        pyconvert(String, metadata.version("matlabengine"))
    catch
        "n/a"
    end
    return (
        julia = string(VERSION),
        pythoncall = _PYTHONCALL_VERSION,
        python = pyconvert(String, sys.executable),
        python_version = pyconvert(String, pyimport("platform").python_version()),
        matlab = pyconvert(String, bridge.module_.matlab_version(bridge.engine)),
        matlabengine = engine_pkg,
        fsr_path = pyconvert(String, bridge.module_.which_fsr(bridge.engine)),
        bridge_dir = bridge.bridge_dir,
        requested_python = bridge.python,
    )
end
